be/src/olap/short_key_index.h

```diff
@@ -0,0 +1,197 @@
+// keys has been written. So we can reserve a certain space for it. which
+// make us must write index file after all indices has been added.
+// index file
+// Usage:
+//      ShortKeyIndexBuilder builder(schema, segment_id, num_rows, delete_flag);
+//      builder.init();
+//      builder.add_item(key1, block1);
+//      ...
+//      builder.add_item(keyN, blockN);
+//      builder.finalize(segment_size, num_rows, &slices);
+class ShortKeyIndexBuilder {
+public:
+    ShortKeyIndexBuilder(const std::vector<FieldInfo>& key_fields,
+                         uint32_t segment_id,
+                         uint32_t num_rows_per_block,
+                         bool delete_flag)
+        : _key_fields(key_fields),
+        _segment_id(segment_id),
+        _num_rows_per_block(num_rows_per_block),
+        _delete_flag(delete_flag) { }
+
+    Status init();
+    // Actually, block_id is useless now, may be removed in some day
+    Status add_item(const RowCursor& key, uint32_t block_id);
+    Status finalize(uint32_t segment_size, uint32_t num_rows, std::vector<Slice>* slices);
+
+private:
+    std::vector<FieldInfo> _key_fields;
+    uint32_t _segment_id = 0;
+    uint32_t _num_rows_per_block;
+    bool _delete_flag;
+
+    // short key length
+    size_t _key_length;
+
+    // used to save serialized key temporarily
+    faststring _tmp_key_buf;
+
+    faststring _header_buf;
+    faststring _index_buf;
+
+    // NOTE: most of these two fields are useless.
+    // put here just for compatible. I will remove this one day
+    // only used to compatible with old version
+    IndexFileHeaderV1 _header;
+};
+
+// Used to decode short key to header and encoded index data.
+// Actually, this class should decode index data to memory format.
+// But for compatible with old code, we reuse MemIndex class.
+// Usage:
+//      MemIndex index;
+//      ShortKeyIndexDecoder decoder(slice)
+//      decoder.parse();
+//      index.load_segment(decoder.header(), decoder.index_data());
+class ShortKeyIndexDecoder {
+public:
+    ShortKeyIndexDecoder(const Slice& data) : _data(data) { }
+    Status parse();
+    const IndexFileHeaderV1& header() const {
+        return _header;
+    }
+    const Slice& index_data() const { return _index_data; }
+private:
+    Slice _data;
+
+    // valid after parse() is called, Header has been removed.
+    // Only contains index data 
+    Slice _index_data;
+
+    // only used to compatible with old version
+    IndexFileHeaderV1 _header;
+};
+
+class ShortKeyIndexIterator {
+public:
+    ShortKeyIndexIterator(const MemIndex* index) : _index(index), _segment(0), _block(0) {
+        _helper.init(_index->short_key_fields());
+    }
+
+    // Position at the first key in the index that at or past target
+    inline void seek_to(const RowCursor& key) {
+        auto offset = _index->find(key, &_helper, false);
+        if (offset.offset > 0) {
+            offset.offset -= 1;
+            offset = _index->next(offset);
```

Comment:
I think here can add comment to explain why to minus one and then call next.

Reply:
OK